test/test_selectors.py

```diff
@@ -0,0 +1,622 @@
+            fd = s.fileno()
+            self.assertTrue(isinstance(fd, int))
+            self.assertGreaterEqual(fd, 0)
+
+    # According to the psutil docs, open_files() has strange behavior
+    # on Windows including giving back incorrect results so to
+    # stop random failures from occurring we're skipping on Windows.
+    @skipIf(sys.platform == "win32", "psutil.Process.open_files() is unstable on Windows.")
+    def test_leaking_fds(self):
+        proc = psutil.Process()
+        before_fds = len(proc.open_files())
+        s = self.make_selector()
+        s.close()
+        after_fds = len(proc.open_files())
+        self.assertEqual(before_fds, after_fds)
+
+    @skipUnless(sys.platform == "win32", "This is the same as test_leaking_fds but for Windows")
+    def test_leaking_fds_windows(self):
+        proc = psutil.Process()
+        before_fds = len(proc.open_files())
+        s = self.make_selector()
+        s.close()
+        after_fds = len(proc.open_files())
+        if before_fds != after_fds:
+            self.skipTest("Leaked fd but test was on Windows.")
+
+class ScalableSelectorMixin:
+    """ Mixin to test selectors that allow more fds than FD_SETSIZE """
+    @skipUnless(resource, "Could not import the resource module")
+    @skipUnless(sys.platform != "darwin", "Can't run on Mac OS due to RINFINITE hard limit.")
+    def test_above_fd_setsize(self):
+        # A scalable implementation should have no problem with more than
+        # FD_SETSIZE file descriptors. Since we don't know the value, we just
+        # try to set the soft RLIMIT_NOFILE to the hard RLIMIT_NOFILE ceiling.
+        soft, hard = resource.getrlimit(resource.RLIMIT_NOFILE)
+
+        try:  # If we're on a *BSD system, the limit tag is different.
+            _, bsd_hard = resource.getrlimit(resource.RLIMIT_OFILE)
+            if bsd_hard < hard:
+                hard = bsd_hard
+
+        # NOTE: AttributeError resource.RLIMIT_OFILE is not defined on Mac OS.
+        except (OSError, resource.error, AttributeError):
+            pass
+
+        try:
+            resource.setrlimit(resource.RLIMIT_NOFILE, (hard, hard))
+            self.addCleanup(resource.setrlimit, resource.RLIMIT_NOFILE,
+                            (soft, hard))
+            limit_nofile = min(hard, 2 ** 16)
+        except (OSError, ValueError):
+            limit_nofile = soft
+
+        # Guard against already allocated FDs
+        limit_nofile -= 256
+        limit_nofile = max(0, limit_nofile)
+
+        s = self.make_selector()
+
+        for i in range(limit_nofile // 2):
+            try:
+                rd, wr = self.make_socketpair()
+            except (OSError, socket.error) as e:
+                # Too many FDs should skip. *BSD and Solaris fail on
+                # connecting or binding rather than on create.
+                if e.errno == errno.EMFILE or e.errno == errno.EADDRNOTAVAIL:
+                    self.skipTest("RLIMIT_NOFILE limit reached.")
```

Comment:
Once again I don't love that we skip the test in this case. I'd argue it should probably be a failure mode.